### DISCUSSION TITLE
How to install?

### DISCUSSION THREAD

**Discussion by saveli** [yu-i-i/overleaf-cep#2] [General]
When trying to build the Dockerfile, I run into this error:
```
user@overleaf:~/overleaf-cep/server-ce$ make build-community
cp .dockerignore ../
docker build \
  --build-arg BUILDKIT_INLINE_CACHE=1 \
  --build-arg OVERLEAF_BASE_TAG \
  --build-arg MONOREPO_REVISION \
  --cache-from sharelatex/sharelatex \
  --cache-from sharelatex/sharelatex:ldap-tc \
  --file Dockerfile \
  --tag sharelatex/sharelatex:ldap-tc-62023e69b3540e87d03550ad64fdab45aa324bf6 \
  --tag sharelatex/sharelatex:ldap-tc \
  ../
[+] Building 0.8s (3/3) FINISHED                                                                            docker:default
 => [internal] load build definition from Dockerfile                                                                  0.0s
 => => transferring dockerfile: 3.83kB                                                                                0.0s
 => [internal] load .dockerignore                                                                                     0.0s
 => => transferring context: 227B                                                                                     0.0s
 => ERROR [internal] load metadata for docker.io/sharelatex/sharelatex-base:ldap-tc-62023e69b3540e87d03550ad64fdab45  0.8s
------
 > [internal] load metadata for docker.io/sharelatex/sharelatex-base:ldap-tc-62023e69b3540e87d03550ad64fdab45aa324bf6:
------
Dockerfile:6
--------------------
   4 |     
   5 |     ARG OVERLEAF_BASE_TAG=sharelatex/sharelatex-base:latest
   6 | >>> FROM $OVERLEAF_BASE_TAG
   7 |     
   8 |     WORKDIR /overleaf
--------------------
ERROR: failed to solve: sharelatex/sharelatex-base:ldap-tc-62023e69b3540e87d03550ad64fdab45aa324bf6: docker.io/sharelatex/sharelatex-base:ldap-tc-62023e69b3540e87d03550ad64fdab45aa324bf6: not found
make: *** [Makefile:30: build-community] Error 1
```
@yu-i-i  How to install?

**Comment by yu-i-i** [comment]
@saveli 
It looks like you havent built base image. Run
make without agruments.

**Reply by saveli** [reply]
That worked so far, but when I enter LDAP details in `variables.env`, I can no longer login. Not with the initial non-LDAP admin account, nor any LDAP account, just this:
![image](https://github.com/user-attachments/assets/8ec9026c-f2e3-4ee3-8242-1c9317baa082)
How to debug?

**Reply by yu-i-i** [reply]
Look at /var/log/overleaf/web.log in the container. What messages are logged there?

**Reply by saveli** [reply]
`{"name":"web","hostname":"c6fb21774dc2","pid":169,"level":50,"err":{"message":"connect ECONNREFUSED <ipadr>:636","name":"Error","stack":"Error: connect ECONNREFUSED <ipadr>:636\n    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1555:16)","info":{},"code":"ECONNREFUSED"},"responseTimeMs":82,"req":{"method":"POST","url":"/login","remoteAddress":"127.0.0.1","headers":{"referer":"https://<ipadr2>/login","user-agent":"Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:128.0) Gecko/20100101 Firefox/128.0","content-length":"80"}},"res":{"statusCode":500,"headers":{"content-length":"796"}},"msg":"POST /login","time":"2024-07-27T13:18:09.179Z","v":0}`

**Reply by saveli** [reply]
Alright, that one was a firewall issue. Confirmed that I can reach ldap server with `ldapsearch` from the docker host.
But now I still have a (different) `ECONNREFUSED`:
When starting with `bin/up -d`:
```
bin/up -d
Initiating Mongo replica set...
[+] Running 1/1
 ✔ Container mongo  Started                                                                                           0.0s 
MongoNetworkError: connect ECONNREFUSED 127.0.0.1:27017
Waiting for Mongo...
[+] Running 4/4
 ✔ Container redis       Started                                                                                      0.0s 
 ✔ Container mongo       Healthy                                                                                      0.0s 
 ✔ Container sharelatex  Started                                                                                      0.0s 
 ✔ Container nginx       Started                                                                                      0.0s 

```

When trying to login:
`{"name":"web","hostname":"c6fb21774dc2","pid":169,"level":50,"err":{"stack":"(no stack)","info":{}},"responseTimeMs":150,"req":{"method":"POST","url":"/login","remoteAddress":"127.0.0.1","headers":{"referer":"https://<ipadr2>/login","user-agent":"Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:128.0) Gecko/20100101 Firefox/128.0","content-length":"105"}},"res":{"statusCode":500,"headers":{"content-length":"796"}},"msg":"POST /login","time":"2024-07-27T14:53:00.324Z","v":0}`

**Reply by yu-i-i** [reply]
@saveli 

Sorry, no idea what's wrong. What I would try to do: add LOG_LEVEL="debug" to variables.env, then:

1) Run upstream image and log in as local user. Does it work? (as far as I understood, you cannot login as local user with ldap version, right?)
2) If it does not work, remove the whole content of 'data' directory, create local admin user and start again (to be sure that ldap version did not do something bad with db. It shouldn't, but nevertheless...)  If it still doesn't work, probably the problem in your configuration.
3) If it does work, try again to run ldap version but in the file  services/web/config/settings.defaults.js comment out 'ldap-authentication' and login as a local user. If it doesn't, what is in the web.log?
4) If it works, uncomment 'ldap-authentication' and try to login as local user, then as ldap user. If it doesn't work, what is in the web.log?

**Reply by saveli** [reply]
@yu-i-i Sadly, debug logging adds a few messages like "first admin account created" and so on, but not more verbosity, so I'm still stuck with "server error 500" when trying to log in.

1. I understand this to remove the docker compose override. I did and I can log in with the local user created under ldap-tc.
2. Removed data a couple of times, unsuccessfully.
3. I added back the docker override, and something with mongodb is different from upstream - now I receive above `ECONNREFUSED` which I did not as long as I only ran ldap-tc with data not removed in between. 
And I get a new error
`{"name":"web","hostname":"9927a04d652c","pid":168,"level":50,"err":{"message":"Cannot read properties of undefined (reading 'enable')","name":"TypeError","stack":"TypeError: Cannot read properties of undefined (reading 'enable')\n    at eval (eval at wrap (/overleaf/node_modules/pug-runtime/wrap.js:6:10), <anonymous>:1021:133)\n    at template (eval at wrap (/overleaf/node_modules/pug-runtime/wrap.js:6:10), <anonymous>:1221:7)\n    at exports.renderFile (/overleaf/node_modules/pug/lib/index.js:454:38)\n    at exports.renderFile (/overleaf/node_modules/pug/lib/index.js:444:21)\n    at exports.__express [as engine] (/overleaf/node_modules/pug/lib/index.js:493:11)\n    at View.render (/overleaf/node_modules/express/lib/view.js:135:8)\n    at tryRender (/overleaf/node_modules/express/lib/application.js:657:10)\n    at Function.render (/overleaf/node_modules/express/lib/application.js:609:3)\n    at ServerResponse.render (/overleaf/node_modules/express/lib/response.js:1048:7)\n    at res.render (/overleaf/services/web/app/src/infrastructure/CSP.js:43:22)","info":{}},"responseTimeMs":7,"req":{"method":"GET","url":"/login","remoteAddress":"127.0.0.1","headers":{"referer":"https://<ipadr2>/login","user-agent":"Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:129.0) Gecko/20100101 Firefox/129.0"}},"res":{"statusCode":500,"headers":{"content-length":"796"}},"msg":"GET /login","time":"2024-08-16T13:57:29.417Z","v":0}`

Website:
![Screenshot from 2024-08-16 15-58-24](https://github.com/user-attachments/assets/51f2e841-38de-4b15-a791-f6758f59a8f5)

4. Above did not work.

Any more ideas?

**Reply by yu-i-i** [reply]
I found a bug when an LDAP user has multiple email addresses. This is probably the issue you're facing. It was fixed in the last commit.

**Reply by saveli** [reply]
Hmm, sadly nothing changed here.

I pulled git changes, ran make again, deleted data, and started the container:
-  I tried with ldap auth enabled to create an admin account with an ldap-email (took the info from a ldapsearch query) and tried to create a local admin. When trying to login with either afterwards, still http 500 error.
-  Then I disabled ldap-auth, deleted data, and ran the docker. Created a local admin, tried to login afterwards, http 500 again.

Basically it is always this trace:
```
"POST /launchpad/register_admin HTTP/1.1" 200
"GET /launchpad HTTP/1.1" 302
"GET /login HTTP/1.1" 200
"GET /js/marketing-0c63cc0df98e15230fec.js HTTP/1.1" 499
"POST /login HTTP/1.1" 500
```

Do I have to additionally limit versions or anything in the docker override?

**Reply by yu-i-i** [reply]
I found a bug that was causing issues with disabled LDAP authentication module. This has been fixed in the latest commits. However, it is unrelated to the 500 error occurring with LDAP authentication.


Could you please after the following line: 
https://github.com/yu-i-i/overleaf-cep/blob/8995efc773f6d5ef3de8a97e3b3141050b6f6dfa/services/web/modules/ldap-authentication/app/src/AuthenticationControllerLdap.js#L30

add
````
console.log('user= ', user)
console.log('ldapUser = ', ldapUser)
````
Then, post the output (with private information redacted).

**Reply by saveli** [reply]
These lines should appear in the browser's console output, right?
Because they don't, so that part is never reached then?

I even added them to the catch part, but still no output:
```
    try {
      user = await AuthenticationManagerLdap.promises.findOrCreateLdapUser(ldapUser, auditLog);
      console.log('user=', user);
      console.log('ldapUser=', ldapUser);
    } catch (error) {
      console.log('user=', user);
      console.log('ldapUser=', ldapUser);
      return {
        user: false,
        info: handleAuthenticateErrors(error, req),
      }
```

**Reply by yu-i-i** [reply]
> These lines should appear in the browser's console output, right?

No, they should appear in the log. Use  
````
overleaf-toolkit/bin/logs -f
````

**Reply by saveli** [reply]
They don't:
```
admin@overleaf:~/toolkit$ bin/logs | grep user=
admin@overleaf:~/toolkit$ bin/logs | grep ldapUser=
admin@overleaf:~/toolkit$ 
```

**Reply by yu-i-i** [reply]
Is the LDAP authentication module enabled in the configuration? If you edit the code in a running container, you need to restart the 'node' process for your changes to take effect.
````
killall -9 node
````
If there is still nothing in the log, after this line:
https://github.com/yu-i-i/overleaf-cep/blob/346ad508e7fee7bb7ba4ef589b426aefae86bbc9/services/web/app/src/Features/Authentication/AuthenticationController.js#L107
output the values of 'err', 'user', and 'infoArray. Try to do this with both an LDAP user and a local user.

**Reply by saveli** [reply]
I did a `git reset --hard` and `git pull`, so ldap was enabled. I checked just now and yes, it is enabled, i.e. not commented out.
Also, I always checked with `docker ps` for running container and used `bin/stop`, `bin/up` respectively, when trying changes.

For local and ldap user:
```
web           | err= unexpected number of matches (41) for "<user>" username
web           | user= undefined
web           | infoArray= undefined
```
User name is irrelevant (email, no email, existing, non-existing), the number 41 doesn't change. I did remove `/data` beforehand.

**Reply by yu-i-i** [reply]
Could you please tell me the values of the variables  `OVERLEAF_LDAP_SEARCH_BASE` and `OVERLEAF_LDAP_SEARCH_FILTER`?  I suspect that the problem in them.

 In the running container, you can find the following code in the file  `/overleaf/node_modules/ldapauth-fork/lib/ldapauth.js` :
````
  switch (result.length) {
    case 0:
      return callback();
    case 1:
      return callback(null, result[0]);
    default:
      return callback(format(
        'unexpected number of matches (%s) for "%s" username',
        result.length, username));
  }
````
From this, we can see that the LDAP search in your case returns 41 results, but the result should be unique or empty. If it is not clear from  `OVERLEAF_LDAP_SEARCH_BASE` and `OVERLEAF_LDAP_SEARCH_FILTER`, you might be able to determine why the search returns so many results by printing them to the log. Start a shell in the container, add `console.log("results = ", results)` before this code, and restart the node processes by running `killall -9 node`

**Reply by saveli** [reply]
Finally it works!

`OVERLEAF_LDAP_SEARCH_FILTER` was configured without the `{{username}}` literal, i.e. the way LDAP usually works - returning a set and then testing the credentials for set inclusion rather than exact match. This part could be highlighted in the readme.md.

Anyhow, many thanks!

**Comment by yu-i-i** [comment]
Sorry, I have no idea why it's not working as expected. I especially don't understand why this is happening:

> I added back the docker override, and something with mongodb is different from upstream - now I receive above ECONNREFUSED which I did not as long as I only ran ldap-tc with data not removed in between.

If 'ldap-authentication' is commented out, the authentication code is from upstream. Just once more. Am I understand right: you create a user in Launchpad with upstream version, then you run ldap-tc  version with commented out  'ldap-authentication', and you cannot authenticate?

**Comment by saveli** [comment]
Solved, see https://github.com/yu-i-i/overleaf-cep/discussions/2#discussioncomment-10844993
